src/io/linux.zig

```diff
@@ -215,9 +300,128 @@ pub const IO = struct {
 
+        if (self.awaiting) |first| {
+            assert(self.awaiting_count > 0);
+            assert(first.awaiting_previous == null);
+            first.awaiting_previous = completion;
+            completion.awaiting_next = first;
+        } else {
+            assert(self.awaiting_count == 0);
+        }
+        self.awaiting = completion;
+        self.awaiting_count += 1;
+
         self.ios_queued += 1;
     }
 
+    /// Cancel should be invoked at most once, before any of the memory owned by read/recv buffers
+    /// is freed (so that lingering async operations do not write to them).
+    ///
+    /// After this function is invoked:
+    /// - No more completion callbacks will be called.
+    /// - No more IO may be submitted.
+    ///
+    /// This function doesn't return until either:
+    /// - All events submitted to io_uring have completed.
+    ///   (They may complete with `error.Canceled`).
+    /// - Or, an io_uring error occurs.
+    ///
+    /// TODO(Linux):
+    /// - Linux kernel ≥5.19 supports the IORING_ASYNC_CANCEL_ALL and IORING_ASYNC_CANCEL_ANY flags,
+    ///   which would allow all events to be cancelled simultaneously with a single "cancel"
+    ///   operation, without IO needing to maintain the `awaiting` doubly-linked list and the `next`
+    ///   cancellation stage.
+    /// - Linux kernel ≥6.0 supports `io_uring_register_sync_cancel` which would remove the `queued`
+    ///   cancellation stage.
+    pub fn cancel(self: *IO) !void {
+        assert(self.cancel_status == .inactive);
+        self.awaiting_verify();
+
+        // Ensure that even if we return early due to an io_uring that IO won't accept more IO.
+        defer self.cancel_status = .done;
+
+        self.cancel_status = .next;
+
+        // Discard any operations that haven't started yet.
+        while (self.unqueued.pop()) |_| {}
+
+        while (self.awaiting) |target| {
+            assert(self.awaiting_count > 0);
+            assert(target.operation != .cancel);
+
+            if (self.cancel_status == .next) {
+                self.cancel_completion = .{
+                    .io = self,
+                    .context = self,
+                    .callback = struct {
+                        fn wrapper(
+                            ctx: ?*anyopaque,
+                            comp: *Completion,
+                            res: *const anyopaque,
+                        ) void {
+                            const io: *IO = @ptrCast(@alignCast(ctx.?));
+                            const result =
+                                @as(*const CancelError!void, @ptrCast(@alignCast(res))).*;
+                            io.cancel_callback(comp, result);
+                        }
+                    }.wrapper,
+                    .operation = .{ .cancel = .{ .target = target } },
+                };
+
+                self.cancel_status = .{ .queued = .{ .target = target } };
+                self.enqueue(&self.cancel_completion);
```

Comment:
maybe extract this code into (non-pub for now) `fn cancel`, and rename current `cancel` to `cancel_all`?